pkg/reconciler/taskrun/taskrun.go

```diff
@@ -603,17 +603,19 @@ func (c *Reconciler) createPod(tr *v1beta1.TaskRun, rtr *resources.ResolvedTaskR
 
 type DeletePod func(podName string, options *metav1.DeleteOptions) error
 
-func updateTaskRunResourceResult(taskRun *v1beta1.TaskRun, podStatus corev1.PodStatus) error {
+func updateTaskRunResourceResult(taskRun *v1beta1.TaskRun, pod corev1.Pod) error {
+	podconvert.SortContainerStatuses(&pod)
+
 	if taskRun.IsSuccessful() {
-		for idx, cs := range podStatus.ContainerStatuses {
+		for idx, cs := range pod.Status.ContainerStatuses {
 			if cs.State.Terminated != nil {
 				msg := cs.State.Terminated.Message
 				r, err := termination.ParseMessage(msg)
 				if err != nil {
 					return fmt.Errorf("parsing message for container status %d: %v", idx, err)
 				}
 				taskResults, pipelineResourceResults := getResults(r)
-				taskRun.Status.TaskRunResults = append(taskRun.Status.TaskRunResults, taskResults...)
+				taskRun.Status.TaskRunResults = appendIfMissingOrReplaced(taskRun.Status.TaskRunResults, taskResults)
```

Comment:
It was a little tricky to follow the logic of `appendIfMissingOrReplaced` for a couple reasons:
- There are multiple nested loops there while we're also inside this container statuses loop. Lots to keep in my head.
- The use of the slice syntax `[:]` to remove duplicates in the newTaskResults list
- Returning the "old" list at the end

Is there an advantage to performing the dedupe process here on every iteration? Could we reduce the complexity a bit by deduplicating just once after iterating through the sorted container statuses?  Something like this:

```go
  uniq := make([]v1beta1.TaskRunResult, 0)
  latest := make(map[string]v1beta1.TaskRunResult, 0)
  for _, res := range taskRun.Status.TaskRunResults {
    if _, seen := latest[res.Name]; !seen {
      uniq = append(uniq, res)
    }
    latest[res.Name] = res
  }
  for i, res := range uniq {
    uniq[i] = latest[res.Name]
  }
  taskRun.Status.TaskRunResults = uniq
```

It's a couple more lines but less nested and (I think) a tad less confusing than the multiple `appends` / slicing / old vs new.

No problem if that's not preferable for some reason, just highlighting a difficulty I had. Maybe rename `appendIfMissingOrReplaced` to `replaceOrAppend`?

Reply:
@sbwsg I like your solution, thanks very much.